linked_list/linked_list.py

```diff
@@ -1,79 +1,173 @@
-        pass
+        pointer = self.head
+        counter = 0
+
+        while pointer != None:
+            counter += 1
+            pointer = pointer.next
+
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: 0(1)
     def get_at_index(self, index):
-        pass
+        if index < 0:
+            return None
+
+        current_index = 0
+        pointer = self.head
+
+        while pointer != None:
+            if current_index == index:
+                return pointer.value
+            pointer = pointer.next
+            current_index += 1
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        # pointer = self.head
+        # next_node = pointer.next
+
+        # while next_node != None:
+        #     pointer = next_node
+        #     next_node = pointer.next
+
+        if self.head is None:
+            return None
+        
+        pointer = self.head
+        
+        while pointer.next is not None:
+            pointer = pointer.next
+            
+        return pointer.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head is None:
+            self.head = new_node
+            return
+        
+        last = self.head
+        while last.next:
+            last = last.next
+
+        last.next = new_node
+        
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head is None:
+            return None
+
+        pointer = self.head
+        maximum = self.head.value
+
+        while pointer is not None:
+            if maximum < pointer.value:
+                maximum = pointer.value
+            pointer = pointer.next
+        
+        return maximum
+
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        position = self.head
+
+        if position is not None:
+            if position.value == value:
+                self.head = position.next
+                position = None
+                return
+
+        while position is not None:
+            if position.value == value:
+                break

```

Comment:
🤓 Generally, we like to avoid break statements where possible. How might you refactor your code to eliminate this one?